proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1232 @@
+
+  subscript(_ position: Int) -> Element { get }
+}
+```
+
+Note that `Span` does _not_ conform to `Collection`. This is because `Collection`, as originally conceived and enshrined in existing source code, assumes pervasive copyability and escapability for itself as well as its elements. In particular a subsequence of a `Collection` is semantically a separate value from the instance it was derived from. In the case of `Span`, a sub-span representing a subrange of its elements _must_ have the same lifetime as the view from which it originates. Another proposal will consider collection-like protocols to accommodate different combinations of `~Copyable` and `~Escapable` for the collection and its elements.
+
+`Span`s representing subsets of consecutive elements can be extracted out of a larger `Span` with an API similar to the recently added `extracting()` functions of `UnsafeBufferPointer`:
+
+```swift
+extension Span {
+  public func extracting(_ bounds: Range<Int>) -> Self
+}
+```
+
+The first element of a given span is _always_ at position zero, and its last it as position `count-1`.
+
+As a side-effect of not conforming to `Collection` or `Sequence`, `Span` is not directly supported by `for` loops at this time. It is, however, easy to use in a `for` loop via indexing:
+
+```swift
+for i in mySpan.indices {
+  calculation(mySpan[i])
+}
+```
+
+### `ContiguousStorage`
+
+A type can declare that it can provide access to contiguous storage by conforming to the `ContiguousStorage` protocol:
+
+```swift
+public protocol ContiguousStorage<Element>: ~Copyable, ~Escapable {
+  associatedtype Element: ~Copyable & ~Escapable
+
+  var storage: Span<Element> { get }
+}
+```
+
+The key safety feature is that a `Span` cannot escape to a scope where the value it borrowed no longer exists.
+
+A function that wishes to read from contiguous storage can declare a parameter type of `some ContiguousStorage`. The implementation will internally consist of a brief generic section, followed by business logic implemented in terms of a concrete `Span`. Frameworks that support library evolution (resilient frameworks) have an additional concern. Resilient frameworks have an ABI boundary that may differ from the API proper. Resilient frameworks may wish to adopt a pattern such as the following:
+
+```swift
+extension MyResilientType {
+  // public API
+  @inlinable public func essentialFunction(_ a: some ContiguousStorage<some Any>) -> Int {
+    self.essentialFunction(a.storage)
+  }
+
+  // ABI boundary
+  public func essentialFunction(_ a: Span<some Any>) -> Int { ... }
+}
+```
+
+Here, the public function obtains the `Span` from the type that vends it in inlinable code, then calls a concrete, opaque function defined in terms of `Span`. Inlining the generic shim in the client is often a critical optimization. The need for such a pattern and related improvements are discussed in the future directions below (see [Syntactic Sugar for Automatic Conversions](#Conversions).)
```

Comment:
For the record, I remain very much unconvinced that it'll be wise to encourage folks to create APIs that can only operate on contiguous storage, as opposed to a series of contiguous storage chunks.